File: examples/go/integ/user_onboarding_test.go
// Copyright (c) 2022-2026 Super Durable, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package integ

import (
	"testing"

	"github.com/stretchr/testify/require"
	"github.com/superdurable/dex/examples/go/products/signup"
	"github.com/superdurable/dex/examples/go/registry"
	"github.com/superdurable/dex/sdk-go/dex"
)

func TestUserOnboardingCompletesEveryTask(t *testing.T) {
	ctx := integrationContext(t)
	flowID := newFlowID(t, "user-onboarding")
	form := signup.SignupForm{
		Username:  flowID,
		Email:     flowID + "@example.com",
		FirstName: "Test",
		LastName:  "User",
	}

	_, err := integClient.StartFlow(
		ctx,
		registry.UserOnboarding,
		flowID,
		form,
		dex.StartFlowOptions{},
	)
	require.NoError(t, err)
	require.NoError(t, integClient.WaitForAttributeEqual(
		ctx,
		flowID,
		signup.Status,
		signup.StatusWaitingForVerification,
	))

	var output string
	require.NoError(t, integClient.InvokeRPC(
		ctx,
		flowID,
		registry.UserOnboarding.Verify,
		nil,
		&output,
		dex.InvokeOptions{},
	))
	require.Equal(t, "verified", output)
	require.NoError(t, integClient.WaitForAttributeEqual(
		ctx,
		flowID,
		signup.Status,
		signup.StatusWaitingForTask1,
	))

	require.NoError(t, integClient.InvokeRPC(
		ctx,
		flowID,
		registry.UserOnboarding.AccomplishTask1,
		nil,
		&output,
		dex.InvokeOptions{},
	))
	require.Equal(t, "task 1 accomplished", output)
	require.NoError(t, integClient.WaitForAttributeEqual(
		ctx,
		flowID,
		signup.Status,
		signup.StatusWaitingForTask2,
	))

	require.NoError(t, integClient.InvokeRPC(
		ctx,
		flowID,
		registry.UserOnboarding.AccomplishTask2,
		nil,
		&output,
		dex.InvokeOptions{},
	))
	require.Equal(t, "task 2 accomplished", output)

	result := waitForFlow(t, flowID)
	require.Equal(t, dex.FlowCompleted, result.Status)
	var completion string
	require.NoError(t, result.Completions[0].Output.Decode(&completion))
	require.Equal(t, "onboarding completed", completion)
}

File: examples/typescript/test/integ/user-onboarding.test.ts
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import assert from "node:assert/strict";
import test from "node:test";

import { stringCodec } from "@superdurable/dex";

import {
  acquireIntegEnvironment,
  releaseIntegEnvironment,
} from "./environment.js";

test.before(async () => {
  await acquireIntegEnvironment();
});

test.after(async () => {
  await releaseIntegEnvironment();
});

test("user onboarding verifies and completes both tasks", async () => {
  const environment = await acquireIntegEnvironment();
  const flow = environment.userOnboardingFlow;
  const flowId = environment.newFlowId("user-onboarding");

  await environment.client.startFlow(
    flow,
    flowId,
    {
      username: flowId,
      email: `${flowId}@example.com`,
      firstName: "Test",
      lastName: "User",
    },
    environment.startOptions(),
  );

  await environment.client.waitForAttributeEqual(
    flowId,
    flow.status,
    "waiting_for_verification",
    20_000,
  );

  assert.equal(await environment.client.invokeRPC(flow.verifySignup, flowId), "verified");
  await environment.client.waitForAttributeEqual(
    flowId,
    flow.status,
    "waiting_for_task_1",
    20_000,
  );
  assert.equal(
    await environment.client.invokeRPC(flow.accomplishTask1, flowId),
    "task 1 accomplished",
  );
  await environment.client.waitForAttributeEqual(
    flowId,
    flow.status,
    "waiting_for_task_2",
    20_000,
  );
  assert.equal(
    await environment.client.invokeRPC(flow.accomplishTask2, flowId),
    "task 2 accomplished",
  );

  const output = await environment.client.waitForFlow(flowId, 45_000).then((result) =>
    result.singleOutput(stringCodec),
  );
  assert.equal(output, "onboarding completed");
});
